sympy/combinatorics/rewritingsystem.py

```diff
@@ -285,3 +289,236 @@ def reduce(self, word, exclude=None):
                 if new != prev:
                     again = True
         return new
+
+    def compute_inverse_rules(self, rules):
+        '''
+        This computes the inverse rules and stores them in the inverse_rules dictionary.
+        The inverse rules are used in the automaton for word reduction.
+
+        '''
+        inverse_rules = {}
+        for r in rules:
+            rule_key_inverse = r**-1
+            rule_value_inverse = (rules[r])**-1
+            if (rule_value_inverse < rule_key_inverse):
+                inverse_rules[rule_key_inverse] = rule_value_inverse
+            else:
+                inverse_rules[rule_value_inverse] = rule_key_inverse
+        return inverse_rules
+
+    def construct_automaton(self):
+        '''
+        Construct the automaton based on the set of reduction rules of the system.
+
+        Automata Design:
+        1. The accept states of the automaton are the proper prefixes of the left hand side of the rules.
+        2. The complete left hand side of the rules are the dead states of the automaton
+
+        '''
+        automaton_alphabet = []
+        left_hand_rules = []
+        proper_prefixes = {}
+
+        # compute all_rules when the automaton is constucted.
+        all_rules = self.rules
+        inverse_rules = self.compute_inverse_rules(all_rules)
+        all_rules.update(inverse_rules)
+
+        generators = list(self.alphabet)
+        generators += [gen**-1 for gen in generators]
+
+        # Contains the alphabets that will be used for state transitions.
+        automaton_alphabet = generators
+
+        # Store the complete left hand side of the rules - dead states.
+        left_hand_rules = list(all_rules)
+
+        # Compute the proper prefixes for every rule.
+        for r in all_rules:
+            proper_prefixes[r] = []
+            letter_word_array = [s for s in r.letter_form_elm]
+            for i in range (1, len(letter_word_array)):
+                letter_word_array[i] = letter_word_array[i-1]*letter_word_array[i]
+            proper_prefixes[r] = letter_word_array
+
+        # Create the states in the automaton.
+        # The left-hand side of the rules are the dead states.
+        # The proper left-hand side of the rules are the accept states.
+
+        # Create a finite state machine as an instance of the StateMachine object
+        fsm = StateMachine('fsm')
+        fsm.add_state('start', is_start=True)
+
+        # Add dead states.
+        for rule in left_hand_rules:
+            if not rule in fsm.states:
+                fsm.add_state(rule, is_dead=True)
```

Comment:
Possibly worth having an `add_states` method instead of `add_state` so that you could add multiple states at once (including just one if need be).

Reply:
Sure 👍